// @pipcook/pipcook-cli/dist/actions/init.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const child_process_1 = require("child_process");
const fs_extra_1 = __importStar(require("fs-extra"));
const inquirer_1 = require("inquirer");
const command_exists_1 = require("command-exists");
const pipcook_core_1 = require("@pipcook/pipcook-core");
const utils_1 = require("../utils");
const config_1 = require("../config");
const { BOA_CONDA_INDEX, BOA_CONDA_MIRROR } = utils_1.Constants;
function npmInstall(npmClient, name, beta, cwd, env) {
    return __awaiter(this, void 0, void 0, function* () {
        child_process_1.execSync(`"${npmClient}" init -f`, { cwd, env, stdio: 'ignore' });
        if (config_1.isLocal) {
            // FIXME(yorkie): manually make symlink to local development
            // This is because npm-install from path will not be compatible with lerna's node_modules/.staging dir.
            const pkg = yield fs_extra_1.readJson(name + '/package.json');
            const dest = path_1.join(cwd, 'node_modules', pkg.name);
            yield fs_extra_1.ensureDir(path_1.dirname(dest));
            yield fs_extra_1.symlink(name, dest, 'dir');
        }
        else {
            if (beta) {
                name = `${name}@beta`;
            }
            const cmd = `"${npmClient}" install ${name} -E --production`;
            console.info(`exec "${cmd}"`);
            child_process_1.execSync(cmd, { cwd, env, stdio: 'inherit' });
        }
    });
}
/**
 * install all dependencies of pipcook into working dir
 */
const init = ({ client, beta, tuna }) => __awaiter(void 0, void 0, void 0, function* () {
    let npmClient = 'npm';
    const npmInstallEnvs = Object.assign({}, process.env);
    if (tuna) {
        npmInstallEnvs.BOA_CONDA_INDEX = BOA_CONDA_INDEX;
        npmInstallEnvs.BOA_CONDA_MIRROR = BOA_CONDA_MIRROR;
        utils_1.logger.info(`switch conda index: ${npmInstallEnvs.BOA_CONDA_INDEX}`);
        utils_1.logger.info(`switch conda mirror: ${npmInstallEnvs.BOA_CONDA_MIRROR}`);
    }
    if (client) {
        npmClient = client;
        if (!config_1.optionalNpmClients.includes(npmClient)) {
            return utils_1.logger.fail(`uknown npm client: ${npmClient}, available clients: ${config_1.optionalNpmClients.join(',')}.`);
        }
    }
    else {
        const clientChoices = [];
        for (const npmClient of config_1.optionalNpmClients) {
            if (command_exists_1.sync(npmClient)) {
                clientChoices.push(npmClient);
            }
        }
        if (clientChoices.length === 1) {
            npmClient = clientChoices[0];
        }
        else if (clientChoices.length > 1) {
            const answer = yield inquirer_1.prompt([
                {
                    type: 'list',
                    name: 'client',
                    message: 'which client do you want to use?',
                    choices: clientChoices
                }
            ]);
            npmClient = answer.client;
        }
        else {
            return utils_1.logger.fail(`no npm client detected`);
        }
    }
    // Install the daemon and pipboard.
    try {
        yield fs_extra_1.default.remove(pipcook_core_1.constants.PIPCOOK_DAEMON);
        yield fs_extra_1.default.remove(pipcook_core_1.constants.PIPCOOK_BOARD);
        yield fs_extra_1.default.ensureDir(pipcook_core_1.constants.PIPCOOK_DAEMON);
        yield fs_extra_1.default.ensureDir(pipcook_core_1.constants.PIPCOOK_BOARD);
        if (tuna) {
            // write the daemon config
            yield fs_extra_1.default.writeJSON(path_1.join(pipcook_core_1.constants.PIPCOOK_HOME_PATH, 'daemon.config.json'), {
                env: {
                    BOA_CONDA_MIRROR
                }
            }, {
                spaces: 2
            });
        }
        yield Promise.all([
            npmInstall(npmClient, config_1.daemonPackage, beta, pipcook_core_1.constants.PIPCOOK_DAEMON, npmInstallEnvs),
            npmInstall(npmClient, config_1.boardPackage, beta, pipcook_core_1.constants.PIPCOOK_BOARD, npmInstallEnvs)
        ]);
        yield fs_extra_1.default.copy(pipcook_core_1.constants.PIPCOOK_BOARD_BUILD, pipcook_core_1.constants.PIPCOOK_DAEMON_PUBLIC);
        utils_1.logger.success('Pipcook is ready, you can try "pipcook --help" to get started.');
    }
    catch (err) {
        utils_1.logger.fail(`failed to initialize Pipcook with the error ${err && err.stack}`, false);
        yield fs_extra_1.default.remove(pipcook_core_1.constants.PIPCOOK_DAEMON);
        yield fs_extra_1.default.remove(pipcook_core_1.constants.PIPCOOK_BOARD);
    }
});
exports.default = init;
